verso/src/lib.rs

```diff
@@ -115,4 +271,179 @@ impl VersoviewController {
             .send(ToVersoMessage::ListenToOnNavigationStarting)?;
         Ok(())
     }
+
+    /// Listen on web resource requests,
+    /// return a boolean in the callback to decide whether or not allowing this navigation
+    pub fn on_web_resource_requested(
+        &self,
+        callback: impl Fn(WebResourceRequest, ResponseFunction) + Send + 'static,
+    ) -> Result<(), Box<ipc_channel::ErrorKind>> {
+        if self
+            .event_listeners
+            .on_web_resource_requested
+            .lock()
+            .unwrap()
+            .replace(Box::new(callback))
+            .is_some()
+        {
+            return Ok(());
+        }
+        self.sender
+            .send(ToVersoMessage::ListenToWebResourceRequests)?;
+        Ok(())
+    }
+
+    /// Sets the webview window's size
+    pub fn set_size<S: Into<Size>>(&self, size: S) -> Result<(), Box<ipc_channel::ErrorKind>> {
+        self.sender.send(ToVersoMessage::SetSize(size.into()))?;
+        Ok(())
+    }
+
+    /// Sets the webview window's position
+    pub fn set_position<P: Into<Position>>(
+        &self,
+        position: P,
+    ) -> Result<(), Box<ipc_channel::ErrorKind>> {
+        self.sender
+            .send(ToVersoMessage::SetPosition(position.into()))?;
+        Ok(())
+    }
+
+    /// Maximize or unmaximize the window
+    pub fn set_maximized(&self, maximized: bool) -> Result<(), Box<ipc_channel::ErrorKind>> {
+        self.sender.send(ToVersoMessage::SetMaximized(maximized))?;
+        Ok(())
+    }
+
+    /// Minimize or unminimize the window
+    pub fn set_minimized(&self, minimized: bool) -> Result<(), Box<ipc_channel::ErrorKind>> {
+        self.sender.send(ToVersoMessage::SetMinimized(minimized))?;
+        Ok(())
+    }
+
+    /// Sets the window to fullscreen or back
+    pub fn set_fullscreen(&self, fullscreen: bool) -> Result<(), Box<ipc_channel::ErrorKind>> {
+        self.sender
+            .send(ToVersoMessage::SetFullscreen(fullscreen))?;
+        Ok(())
+    }
+
+    /// Show or hide the window
+    pub fn set_visible(&self, visible: bool) -> Result<(), Box<ipc_channel::ErrorKind>> {
+        self.sender.send(ToVersoMessage::SetVisible(visible))?;
+        Ok(())
+    }
+
+    /// Moves the window with the left mouse button until the button is released
+    pub fn start_dragging(&self) -> Result<(), Box<ipc_channel::ErrorKind>> {
+        self.sender.send(ToVersoMessage::StartDragging)?;
+        Ok(())
+    }
+
+    /// Get the window's size
+    pub fn get_size(&self) -> Result<PhysicalSize<u32>, Box<ipc_channel::ErrorKind>> {
+        self.sender.send(ToVersoMessage::GetSize)?;
+        let (sender, receiver) = std::sync::mpsc::channel();
+        self.event_listeners
+            .size_response
+            .lock()
+            .unwrap()
+            .replace(sender);
+        Ok(receiver.recv().unwrap())
+    }
```

Comment:
Things like these work like this:

1. we send `ToVersoMessage::GetSize` here, and wait with a blocking receive
2. versoview receives this message and push send us back a `ToControllerMessage::GetSizeResponse`
3. We receive this message from the message look created in the beginning
4. Our message look send the message to us
5. We receive this message and return it to the caller

Reply:
hmm, I think at least we could handle the race condition between receiving messages from Verso and setting up the channel.

```rust
self.sender.send(ToVersoMessage::GetSize)?;
// <--- Verso may response ToControllerMessage::GetSizeResponse here
let (sender, receiver) = std::sync::mpsc::channel();
// <--- Verso may response ToControllerMessage::GetSizeResponse here
self.event_listeners
    .size_response
    .lock()
    .unwrap()
    .replace(sender);
```

Reply:
Ah nice catch, didn't think of this, have changed it to something like this

```rust
```